acceptance: TestDockerCLI failed

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2021716&tab=buildLog) on [master@a523cf3564e00275d990fbf14f49e84c0d6be400](https://github.com/cockroachdb/cockroach/commits/a523cf3564e00275d990fbf14f49e84c0d6be400):

```
=== RUN   TestDockerCLI
--- FAIL: TestDockerCLI (319.76s)
    test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI453901509
    test_log_scope.go:58: use -show-logs to present logs inline
=== RUN   TestDockerCLI/test_demo_partitioning.tcl
    --- FAIL: TestDockerCLI/test_demo_partitioning.tcl (43.85s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2053136&tab=buildLog) on [master@a6a58f4bd1d6b9a7d9f0de2c690d35b51d4f41ca](https://github.com/cockroachdb/cockroach/commits/a6a58f4bd1d6b9a7d9f0de2c690d35b51d4f41ca):

```
=== RUN   TestDockerCLI
    TestDockerCLI: test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI038365640
    TestDockerCLI: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestDockerCLI (367.99s)
=== RUN   TestDockerCLI/test_demo_partitioning.tcl
    --- FAIL: TestDockerCLI/test_demo_partitioning.tcl (44.35s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2053293&tab=buildLog) on [master@a6a58f4bd1d6b9a7d9f0de2c690d35b51d4f41ca](https://github.com/cockroachdb/cockroach/commits/a6a58f4bd1d6b9a7d9f0de2c690d35b51d4f41ca):

```
=== RUN   TestDockerCLI
    TestDockerCLI: test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI062369959
    TestDockerCLI: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestDockerCLI (354.01s)
=== RUN   TestDockerCLI/test_demo_partitioning.tcl
    --- FAIL: TestDockerCLI/test_demo_partitioning.tcl (45.81s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2056119&tab=buildLog) on [master@e99e7a8c2d40db3e394eef8042e3d122d1401498](https://github.com/cockroachdb/cockroach/commits/e99e7a8c2d40db3e394eef8042e3d122d1401498):

```
=== RUN   TestDockerCLI
    TestDockerCLI: test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI286518199
    TestDockerCLI: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestDockerCLI (350.01s)
=== RUN   TestDockerCLI/test_high_verbosity.tcl
    --- FAIL: TestDockerCLI/test_high_verbosity.tcl (31.47s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2056305&tab=buildLog) on [master@51807b66702f5cff9af9a18fe7851609c7b79a68](https://github.com/cockroachdb/cockroach/commits/51807b66702f5cff9af9a18fe7851609c7b79a68):

```
=== RUN   TestDockerCLI
    TestDockerCLI: test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI518002461
    TestDockerCLI: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestDockerCLI (354.65s)
=== RUN   TestDockerCLI/test_demo_partitioning.tcl
    --- FAIL: TestDockerCLI/test_demo_partitioning.tcl (44.58s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(acceptance).TestDockerCLI failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2057755&tab=buildLog) on [master@415e2bcff3f07e7fc2130233bf6443f61ed51cb9](https://github.com/cockroachdb/cockroach/commits/415e2bcff3f07e7fc2130233bf6443f61ed51cb9):

```
=== RUN   TestDockerCLI
    TestDockerCLI: test_log_scope.go:77: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDockerCLI220337452
    TestDockerCLI: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestDockerCLI (352.03s)
=== RUN   TestDockerCLI/test_high_verbosity.tcl
    --- FAIL: TestDockerCLI/test_high_verbosity.tcl (32.60s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDockerCLI PKG=./pkg/acceptance TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDockerCLI.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/storage/engine.go
+++ b/pkg/storage/engine.go
@@ -64,10 +64,10 @@ type SimpleIterator interface {
 	// for a key.
 	NextKey()
 	// UnsafeKey returns the same value as Key, but the memory is invalidated on
-	// the next call to {Next,Prev,Seek,SeekReverse,Close}.
+	// the next call to {Next,NextKey,Prev,SeekGE,SeekLT,Close}.
 	UnsafeKey() MVCCKey
 	// UnsafeValue returns the same value as Value, but the memory is
-	// invalidated on the next call to {Next,Prev,Seek,SeekReverse,Close}.
+	// invalidated on the next call to {Next,NextKey,Prev,SeekGE,SeekLT,Close}.
 	UnsafeValue() []byte
 }
 

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -1322,6 +1322,7 @@ func replayTransactionalWrite(
 ) error {
 	var found bool
 	var writtenValue []byte
+	var writtenValueSafety valueSafety
 	var err error
 	metaKey := MakeMVCCMetadataKey(key)
 	if txn.Sequence == meta.Txn.Sequence {
@@ -1332,7 +1333,7 @@ func replayTransactionalWrite(
 		defer getBuf.release()
 		getBuf.meta = buf.meta
 		var exVal *roachpb.Value
-		if exVal, _, _, err = mvccGetInternal(
+		if exVal, _, writtenValueSafety, err = mvccGetInternal(
 			ctx, iter, metaKey, timestamp, true /* consistent */, unsafeValue, txn, getBuf); err != nil {
 			return err
 		}
@@ -1341,6 +1342,7 @@ func replayTransactionalWrite(
 	} else {
 		// Get the value from the intent history.
 		writtenValue, found = meta.GetIntentValue(txn.Sequence)
+		writtenValueSafety = safeValue
 	}
 	if !found {
 		return errors.Errorf("transaction %s with sequence %d missing an intent with lower sequence %d",
@@ -1360,8 +1362,14 @@ func replayTransactionalWrite(
 			prevVal := prevIntent.Value
 
 			exVal = &roachpb.Value{RawBytes: prevVal}
-		}
-		if exVal == nil {
+		} else {
+			// We're going to be using the iterator again, so make sure the
+			// existing writtenValue bytes are safe and won't be corrupted.
+			if writtenValueSafety == unsafeValue {
+				writtenValue = append([]byte(nil), writtenValue...)
+				writtenValueSafety = safeValue
+			}
+
 			// If the previous value at the key wasn't written by this
 			// transaction, or it was hidden by a rolled back seqnum, we
 			// look at last committed value on the key.

--- a/pkg/storage/testdata/mvcc_histories/idempotent_transactions
+++ b/pkg/storage/testdata/mvcc_histories/idempotent_transactions
@@ -54,7 +54,7 @@ error: (*withstack.withStack:) transaction 00000000-0000-0000-0000-000000000001 
 
 run ok
 with t=a k=i
-  # On a separate key, start an increment.
+  # On a separate nonexistent key, start an increment.
   txn_step n=3
   increment
   # As long as the sequence is unchanged, replaying the increment doesn't increase the value.
@@ -75,8 +75,11 @@ data: "i"/0.000000011,0 -> /INT/1
 
 run ok
 with t=a k=i
-  # Increment again.
+  # Increment previously nonexistent key again.
   txn_step
+  increment
+  # As long as the sequence is unchanged, replaying the increment doesn't increase the value.
+  increment
   increment
   txn_step n=-1
   # Replaying an older increment doesn't increase the value.
@@ -84,6 +87,8 @@ with t=a k=i
   increment
   increment
 ----
+inc: current value = 2
+inc: current value = 2
 inc: current value = 2
 inc: current value = 1
 inc: current value = 1
@@ -94,3 +99,69 @@ meta: "a"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 
 data: "a"/0.000000011,0 -> /BYTES/second
 meta: "i"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=3} ts=0.000000011,0 del=false klen=12 vlen=6 ih={{2 /INT/1}}
 data: "i"/0.000000011,0 -> /INT/2
+
+# Write a key outside of the transaction.
+run ok
+increment k=i2 ts=10
+----
+inc: current value = 1
+>> at end:
+meta: "a"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=1} ts=0.000000011,0 del=false klen=12 vlen=11 ih={{0 /BYTES/first}}
+data: "a"/0.000000011,0 -> /BYTES/second
+meta: "i"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=3} ts=0.000000011,0 del=false klen=12 vlen=6 ih={{2 /INT/1}}
+data: "i"/0.000000011,0 -> /INT/2
+data: "i2"/0.000000010,0 -> /INT/1
+
+run ok
+with t=a k=i2
+  # On a separate existent key, start an increment.
+  txn_step n=2
+  increment
+  # As long as the sequence is unchanged, replaying the increment doesn't increase the value.
+  increment
+  increment
+  increment
+----
+inc: current value = 2
+inc: current value = 2
+inc: current value = 2
+inc: current value = 2
+>> at end:
+txn: "a" meta={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=4} lock=true stat=PENDING rts=0.000000011,0 wto=false max=0,0
+meta: "a"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=1} ts=0.000000011,0 del=false klen=12 vlen=11 ih={{0 /BYTES/first}}
+data: "a"/0.000000011,0 -> /BYTES/second
+meta: "i"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=3} ts=0.000000011,0 del=false klen=12 vlen=6 ih={{2 /INT/1}}
+data: "i"/0.000000011,0 -> /INT/2
+meta: "i2"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=4} ts=0.000000011,0 del=false klen=12 vlen=6
+data: "i2"/0.000000011,0 -> /INT/2
+data: "i2"/0.000000010,0 -> /INT/1
+
+run ok
+with t=a k=i2
+  # Increment previously existent key again.
+  txn_step
+  increment
+  # As long as the sequence is unchanged, replaying the increment doesn't increase the value.
+  increment
+  increment
+  txn_step n=-1
+  # Replaying an older increment doesn't increase the value.
+  increment
+  increment
+  increment
+----
+inc: current value = 3
+inc: current value = 3
+inc: current value = 3
+inc: current value = 2
+inc: current value = 2
+inc: current value = 2
+>> at end:
+txn: "a" meta={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=4} lock=true stat=PENDING rts=0.000000011,0 wto=false max=0,0
+meta: "a"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=1} ts=0.000000011,0 del=false klen=12 vlen=11 ih={{0 /BYTES/first}}
+data: "a"/0.000000011,0 -> /BYTES/second
+meta: "i"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=3} ts=0.000000011,0 del=false klen=12 vlen=6 ih={{2 /INT/1}}
+data: "i"/0.000000011,0 -> /INT/2
+meta: "i2"/0,0 -> txn={id=00000000 key="a" pri=0.00000000 epo=0 ts=0.000000011,0 min=0,0 seq=5} ts=0.000000011,0 del=false klen=12 vlen=6 ih={{4 /INT/2}}
+data: "i2"/0.000000011,0 -> /INT/3
+data: "i2"/0.000000010,0 -> /INT/1

